astrokat/noisediode.py

```diff
@@ -101,16 +90,17 @@ def _set_dig_nd_(kat,
 
 def _katcp_reply_(dig_katcp_replies):
     """ KATCP timestamp return logs"""
+    ant_ts_list = []
     for ant in sorted(dig_katcp_replies):
         reply, informs = dig_katcp_replies[ant]
-        # test incorrect reply check
-        if len(reply.arguments) < 2:
-            msg = 'Unexpected response after noise diode instruction'
-            user_logger.warn(msg.format(ant))
+        if reply.succeeded():
```

Comment:
Sorry,  I made a mistake with this suggestion.  In this case it needs to be `reply.reply_ok()`.
For added safety, we could include `len(reply.arguments) >= 4` (that's how many `_nd_log_msg_` needs).

There's a subtle difference depending on whether the return value from the KATCP request is unpacked into reply and informs, or just left as a single object:

```python
In [14]: response = cam.m011.req.dig_capture_list()

In [15]: reply, informs  = cam.m011.req.dig_capture_list()

In [16]: type(response)
Out[16]: katcp.resource.KATCPReply

In [17]: type(reply)
Out[17]: katcp.core.Message

In [18]: response.succeeded
Out[18]: True

In [19]: reply.succeeded
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-19-d57bc05188ca> in <module>()
----> 1 reply.succeeded

AttributeError: 'Message' object has no attribute 'succeeded'

In [20]: reply.reply_ok()
Out[20]: True
```

Reply:
Tx @ajoubertza I did not know how to test this on my laptop
I update to `reply.reply_ok()` in `_katcp_reply_`

The safety check for the number of arguments I implemented in `_nd_log_msg_`, just in front of where it is used

```

    if len(reply.arguments) < 4:
        msg = 'Unexpected number of return arguments\n'
        msg += '4 values expected: code time on-frac cycle-len'
        msg += 'Found {}'.format(len(reply.arguments))
        raise RuntimeError(msg)

    actual_time = float(reply.arguments[1])
    actual_on_frac = float(reply.arguments[2])
    actual_cycle = float(reply.arguments[3])
```

Reply:
I also agree that we should retain the `len` evaluation.
I found the reason I put it in there -- to avoid the obvious `IndexError` exception
```
2019-07-28 16:15:32.678Z INFO     Switch noise-diode off at 1564330538.0
Unhandled exception <type 'exceptions.IndexError'> : list index out of range
Traceback (most recent call last):
  File "/usr/local/bin/astrokat-observe.py", line 7, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/usr/local/lib/python2.7/dist-packages/astrokat/observe_main.py", line 670, in main
    run_observation(opts, kat)
  File "/usr/local/lib/python2.7/dist-packages/astrokat/observe_main.py", line 259, in __exit__
    noisediode.off(self.array)
  File "/usr/local/lib/python2.7/dist-packages/astrokat/noisediode.py", line 115, in off
    _katcp_reply_to_log_(replies)
  File "/usr/local/lib/python2.7/dist-packages/astrokat/noisediode.py", line 19, in _katcp_reply_to_log_
    _nd_log_msg_(ant, reply, informs)
  File "/usr/local/lib/python2.7/dist-packages/astrokat/noisediode.py", line 34, in _nd_log_msg_
    actual_time = float(reply.arguments[1])
Running cleanup for sched
Running cleanup for katpool
Task 20190727-0014 entered state 'FAILED' with reason: Non-zero script process exit status: 1.
```